solutions/cf-839C.mdx

```diff
@@ -0,0 +1,106 @@
+---
+id: cf-839C
+source: CF
+title: Journey
+author: Reyansh Bahl
+---
+
+[Official Solution (C++)](https://codeforces.com/blog/entry/53815)
+
+## Explanation
+
+The first step is to calculate the length and probability of each possible path.
+We can do this using DFS. We start of with a probability of 1, or 100%. Then,
+for each node, we divide the current path's probability by the number of
+possible next moves.
+
+For example, if node 1 is connected to nodes 2 and 3, the probability of
+reaching node 2 is $1/2$, or $0.5$, and the probability of reaching node 3 is
+also $0.5$. If node 3 is connected to nodes 4 and 5, the probability of reaching
+node 4 is $0.5 / 2 = 0.25$ and the probability of reaching node 5 is also
+$0.25$. Let's say that nodes 2, 4, and 5 are leaf nodes. The possible paths are
+listed below:
+
+- $1 \rightarrow 2$ : 0.5
+- $1 \rightarrow 3 \rightarrow 4$ : 0.25
+- $1 \rightarrow 3 \rightarrow 5$ : 0.25
+
+To calculate the expected length of the journey, we multiply each of the
+probabilities with the corresponding path length. In the example above, the
+expected length would be $1\cdot0.5 + 2\cdot0.25 + 2\cdot0.25 = 1.5$
+
+In the code, for each node, we calculate the number of possible next moves by
+looping through the node's neighbors and counting how many are unvisited. If
+there are no such nodes, then the path has ended and we can add the length of
+the path times the probability to the final answer. Otherwise, we divide the
+probability (as described above) and continue the DFS.
+
+Note: `setprecision(10)` sets the number of decimal places to be printed to be 10. See [this](https://cplusplus.com/reference/iomanip/setprecision/) for more details.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int n;
+vector<vector<int>> adj;  // adjacency list
+vector<bool> visited;  // stores which nodes have been visited
+
+long double ans;
+
+void dfs(int node, int current_length, double current_probability) {
+	visited[node] = true;
+	int possible_moves = 0;
+	// count the number of cities we can move to
+	for (int x : adj[node]) {
+		if (!visited[x]) {
+			possible_moves++;
+		}
+	}
+
+	if (!possible_moves) {
+		// if there are no cities to move to, the path has ended
+		ans += current_length * current_probability;
+	} else {
+		// update the new probability by dividing it
+		// by the number of possible cities
+		double new_probability = current_probability / possible_moves;
+		for (int x : adj[node]) {
+			if (!visited[x]) {
+				dfs(x, current_length + 1, new_probability);
+			}
+		}
+	}
+
+	// so that this node can be part of other possible paths
```

Comment:
explain this more / clarify because this can be a bit confusing

Reply:
The code actually was unnecessary since there aren't any cycles, so I removed it. Thanks.